segments/gitstatus/gitstatus.p9k

```diff
@@ -0,0 +1,146 @@
+  #                                                                                                                              
+  p9k::register_icon "GITSTATUS_GITHUB"        ''          $'\uE20E '  $'\uF113 '  '\u'${CODEPOINT_OF_AWESOME_GITHUB_ALT}' '         $'\uF113 '
+  #                                                                                                                              
+  p9k::register_icon "GITSTATUS_BITBUCKET"     ''          $'\uE20E '  $'\uF171 '  '\u'${CODEPOINT_OF_AWESOME_BITBUCKET}' '          $'\uE703 '
+  #                                                                                                                              
+  p9k::register_icon "GITSTATUS_GITLAB"        ''          $'\uE20E '  $'\uF296 '  '\u'${CODEPOINT_OF_AWESOME_GITLAB}' '             $'\uF296 '
+
+  ################################################################
+  # Define the actual segment
+  ################################################################
+
+  ################################################################
+  # Register segment
+  # Parameters:
+  #   segment_name  context  background  foreground  Generic  Flat/Awesome-Patched  Awesome-FontConfig  Awesome-Mapped-FontConfig  NerdFont
+  p9k::register_segment  "GITSTATUS" "CLEAN"     "green"  "${DEFAULT_COLOR}" ''  ''  ''  ''  ''
+  p9k::register_segment  "GITSTATUS" "MODIFIED"  "yellow" "${DEFAULT_COLOR}" ''  ''  ''  ''  ''
+  p9k::register_segment  "GITSTATUS" "UNTRACKED" "green"  "${DEFAULT_COLOR}" '?' $'\uE16C' $'\uF059' '\u'${CODEPOINT_OF_AWESOME_QUESTION_CIRCLE} $'\uF059'
+
+  ################################################################
+  # Register segment default values
+  p9k::set_default P9K_GITSTATUS_MAX_INDEX_SIZE_DIRTY "4096"
+  p9k::set_default P9K_GITSTATUS_MAX_SYNC_LATENCY_SECONDS "5"
+  p9k::set_default P9K_GITSTATUS_SHOW_CHANGESET false
+  p9k::set_default P9K_GITSTATUS_CHANGESET_HASH_LENGTH 8
+  p9k::set_default P9K_GITSTATUS_ALWAYS_SHOW_REMOTE_BRANCH false
+  p9k::set_default P9K_GITSTATUS_HIDE_BRANCH_ICON false
+  p9k::set_default P9K_GITSTATUS_ACTIONFORMAT_FOREGROUND "001"
+}
+
+## Start the gitstatus daemon
+source ${P9K_GITSTATUS_DIR:-${_P9K_INSTALLATION_DIR}/gitstatus}/gitstatus.plugin.zsh
+gitstatus_start -m ${P9K_GITSTATUS_MAX_INDEX_SIZE_DIRTY} P9K
+
+################################################################
+# @description
+#   Displays the status of your git checkout.
+##
+# @args
+#   $1 string Alignment - left | right
+#   $2 integer Segment index
+#   $3 boolean Whether the segment should be joined
+##
+prompt_gitstatus() {
+  gitstatus_query -d ${${GIT_DIR:a}:-$PWD} -t $P9K_GITSTATUS_MAX_SYNC_LATENCY_SECONDS P9K 2>/dev/null
+
+  # Print nothing, if we are not in a repo, or if the gitstatus daemon sends a timeout
+  [[ -z "${VCS_STATUS_RESULT}" || "${VCS_STATUS_RESULT}" =~ "norepo-" || "${VCS_STATUS_RESULT}" == "tmout" ]] && return
```

Comment:
You mean a negated check for `ok-*`? This is the early exit here..

Reply:
Yes, of course.

```zsh
[[ $VCS_STATUS_RESULT =~ ok-* ]] || return
```

Reply:
Btw. that would work, except if gitstatus returns a timeout. I am a bit undecided, what I should do in that state, TBH. With the current implementation it shows (wrongly) a detached state (clarification: the segment code does this, not gitstatus), which is maybe better than showing nothing. On the other hand I could handle the timeout case differently (showing something like "timeout", or similar..).

Reply:
There are a few options.

The simplest is to just disable timeout. Since `vcs_info` is invoked without a timeout, it kinda makes sense. On the other hand, gitstatus is way less stable than `vcs_info` (by about the same factor it's faster, which is to say a lot less stable). In addition, unlike `vcs_info`, crashes in `gitstatus` will lead to hangs rather than simple errors. All-in-all, I don't recommend going without a timeout.

You could fall back to `vcs_info`. But given that it's you've implemented gitstatus as a separate segment, it makes little sense.

You could pretend it's not a git repo. This will look smooth but can be misleading.

You could display "error" in the segment, perhaps with a hint to blame gitstatus so that people will file bugs directly against me rather bothering you. So, "gitstatus error". I think this is the best option.

Reply:
Okay. Thanks for your input. I output `gitstatus timeout` now, if there is a timeout. Additionally, there is a section for troubleshooting in the README (with the hint to increase the timeout). Think that should be enough, until we get to the point being able to fall back to async.

Reply:
That's perfect.